sdk/storage/azblob/internal/exported/blob_batch.go

```diff
@@ -0,0 +1,269 @@
+	"io"
+	"io/ioutil"
+	"mime/multipart"
+	"net/http"
+	"net/textproto"
+	"strconv"
+	"strings"
+)
+
+type BlobBatchOperationType string
+
+const (
+	BatchIdPrefix                                    = "batch_"
+	HttpVersion                                      = "HTTP/1.1"
+	HttpNewline                                      = "\r\n"
+	BatchDeleteOperationType  BlobBatchOperationType = "delete"
+	BatchSetTierOperationType BlobBatchOperationType = "set tier"
+)
+
+type BlobBatchBuilder struct {
+	Endpoint    *string
+	AuthPolicy  policy.Policy
+	SubRequests []*policy.Request
+}
+
+// createBatchID is used for creating a new batch id which is used as batch boundary in the request body
+func createBatchID() (string, error) {
+	batchID, err := uuid.New()
+	if err != nil {
+		return "", err
+	}
+
+	return BatchIdPrefix + batchID.String(), nil
+}
+
+// buildSubRequest is used for building the sub-request. Example:
+// DELETE /container0/blob0 HTTP/1.1
+// x-ms-date: Thu, 14 Jun 2018 16:46:54 GMT
+// Authorization: SharedKey account:G4jjBXA7LI/RnWKIOQ8i9xH4p76pAQ+4Fs4R1VxasaE=
+// Content-Length: 0
+func buildSubRequest(req *policy.Request) []byte {
+	var batchSubRequest strings.Builder
+	blobPath := req.Raw().URL.Path
+	if len(req.Raw().URL.RawQuery) > 0 {
+		blobPath += "?" + req.Raw().URL.RawQuery
+	}
+
+	batchSubRequest.WriteString(fmt.Sprintf("%s %s %s%s", req.Raw().Method, blobPath, HttpVersion, HttpNewline))
+
+	for k, v := range req.Raw().Header {
+		if strings.EqualFold(k, shared.HeaderXmsVersion) {
+			continue
+		}
+		if len(v) > 0 {
+			batchSubRequest.WriteString(fmt.Sprintf("%v: %v%v", k, v[0], HttpNewline))
+		}
+	}
+
+	batchSubRequest.WriteString(HttpNewline)
+	return []byte(batchSubRequest.String())
+}
+
+// CreateBatchRequest creates a new batch request using the sub-requests present in the BlobBatchBuilder.
+//
+// Example of a sub-request in the batch request body:
+//
+//	--batch_357de4f7-6d0b-4e02-8cd2-6361411a9525
+//	Content-Type: application/http
+//	Content-Transfer-Encoding: binary
+//	Content-ID: 0
+//
+//	DELETE /container0/blob0 HTTP/1.1
+//	x-ms-date: Thu, 14 Jun 2018 16:46:54 GMT
+//	Authorization: SharedKey account:G4jjBXA7LI/RnWKIOQ8i9xH4p76pAQ+4Fs4R1VxasaE=
+//	Content-Length: 0
+func CreateBatchRequest(bb *BlobBatchBuilder) ([]byte, string, error) {
+	batchID, err := createBatchID()
+	if err != nil {
+		return nil, "", err
+	}
+
+	// Create a new multipart buffer
+	reqBody := &bytes.Buffer{}
+	writer := multipart.NewWriter(reqBody)
+
+	// Set the boundary
+	err = writer.SetBoundary(batchID)
+	if err != nil {
+		return nil, "", err
+	}
+
+	partHeaders := make(textproto.MIMEHeader)
+	partHeaders["Content-Type"] = []string{"application/http"}
+	partHeaders["Content-Transfer-Encoding"] = []string{"binary"}
+	var partWriter io.Writer
+
+	for i, req := range bb.SubRequests {
+		if bb.AuthPolicy != nil {
+			resp, err := bb.AuthPolicy.Do(req)
+			if err != nil && resp != nil {
```

Comment:
I don't think this is correct as the returned `resp` and `err` are mutually exclusive.

Also, if `err` isn't `nil`, shouldn't the error be returned here?

Reply:
The auth policy is just executed here which has the following code: `res, err := req.Next()`
The `Next()` method has this check,
```
if len(req.policies) == 0 {
	return nil, errors.New("no more policies")
}
```
So, `bb.AuthPolicy.Do(req)` will always return the error `errors.New("no more policies")`, since the auth policy is directly executed here.

We can update the logic to do something like this,
```
if err != nil && !errors.Is(err, errors.New("no more policies")) {
	if log.Should(EventSubmitBatch) {
		log.Writef(EventSubmitBatch, "failed to authorize sub-request for %v.\nError: %v", req.Raw().URL.Path, err.Error())
	}
}
```

@jhendrixMSFT , should we fail the entire batch here if the auth policy is failing for a single sub-request?

Reply:
I think it should fail.  The reason is because in the singular case, if the auth policy fails the request fails, and I think having this be consistent is good.